gateway-provider-security-jwt/src/main/java/org/apache/knox/gateway/provider/federation/jwt/filter/AccessTokenFederationFilter.java

```diff
@@ -176,6 +172,6 @@ private Subject createSubjectFromToken(JWTToken token) {
     // To modify the Principals Set, the caller must have AuthPermission("modifyPrincipals").
     // To modify the public credential Set, the caller must have AuthPermission("modifyPublicCredentials").
     // To modify the private credential Set, the caller must have AuthPermission("modifyPrivateCredentials").
-    return new javax.security.auth.Subject(true, principals, emptySet, emptySet);
+    return new javax.security.auth.Subject(true, principals, Collections.emptySet(), Collections.emptySet());
```

Comment:
LGTM with one note. 
I'm not sure if it's a real problem, but since we're using `Collections.emptySet()` here, this means that adding a new principal after this point (e.g.: `subject.getPrincipals().add()`) to the subject might fail because the `Collection.emptySet()` is unmodifiable. Unlike the `new HashSet<>();`.

Reply:
This is a valid question. The answer is that it's even better than the implementation before due to the very same reason you described: once this subject is created, the principals cannot be modified. In lower layers, like in identity assertion, if someone wants to proceed with different principals, a new Subject has to be created and call `Subject.doAs`.